src/Dialect/ONNX/Transforms/LegalizeQuarkQuantizedOps.cpp

```diff
@@ -0,0 +1,533 @@
+#include <mlir/IR/BuiltinTypes.h>
+#include <mlir/IR/Diagnostics.h>
+#include <mlir/IR/Dialect.h>
+#include <mlir/IR/IRMapping.h>
+#include <mlir/IR/MLIRContext.h>
+#include <mlir/IR/Matchers.h>
+#include <mlir/IR/Operation.h>
+#include <mlir/IR/PatternMatch.h>
+#include <mlir/IR/Value.h>
+#include <mlir/IR/Verifier.h>
+#include <mlir/Pass/Pass.h>
+#include <mlir/Pass/PassManager.h>
+#include <mlir/Support/LogicalResult.h>
+#include <mlir/Transforms/DialectConversion.h>
+#include <mlir/Transforms/GreedyPatternRewriteDriver.h>
+
+#include "src/Compiler/CompilerOptions.hpp"
+#include "src/Dialect/ONNX/DialectBuilder.hpp"
+#include "src/Dialect/ONNX/ElementsAttr/ElementsAttrHelper.hpp"
+#include "src/Dialect/ONNX/ONNXDialect.hpp"
+#include "src/Dialect/ONNX/ONNXOps.hpp"
+#include "src/Dialect/ONNX/ONNXOps/OpHelper.hpp"
+#include "src/Dialect/ONNX/Transforms/LegalizeQuarkQuantizedOps.hpp"
+#include "src/Pass/Passes.hpp"
+
+using namespace mlir;
+
+namespace {
+
+Operation *createOpWithNewType(PatternRewriter &rewriter, Operation *op,
+    SmallVector<Value> &operands, llvm::SmallVector<Location> &locs,
+    llvm::SmallVector<Type> &types, const FloatType toFloatType) {
+
+  auto fusedLoc = FusedLoc::get(op->getContext(), locs);
+
+  // Requires special treatment for ONNXCastOp as we need to modify its 'to'
+  // attribute.
+  if (auto onnxCastOp = dyn_cast<mlir::ONNXCastOp>(op)) {
+    assert(operands.size() == 1 && "CastOp must have a simple operand");
+    auto inputValue = operands.back();
+    auto onnxCastTy = cast<mlir::TensorType>(inputValue.getType());
+    auto newOnnxCastTy = onnxCastTy.clone(toFloatType);
+    return rewriter.create<mlir::ONNXCastOp>(fusedLoc, newOnnxCastTy,
+        operands.back(), onnxCastOp.getSaturate(), toFloatType);
+  }
+
+  // Start composing new op
+  OperationState state(fusedLoc, op->getName().getStringRef(), operands, types,
+      op->getAttrs(), op->getSuccessors());
+
+  // Create the new op
+  rewriter.setInsertionPoint(op);
+  return rewriter.create(state);
+}
+
+bool isQuarkGeneratedCastOp(
+    Operation *op, const FloatType fromFloatType, const FloatType toFloatType) {
+  if (!isa_and_nonnull<mlir::ONNXCastOp>(op)) {
+    return false;
+  }
+  auto castOp = cast<mlir::ONNXCastOp>(op);
+  return castOp.getTo() == toFloatType &&
+         cast<TensorType>(castOp.getInput().getType()).getElementType() ==
+             fromFloatType;
+}
+
+bool isInputCastOp(
+    Operation *op, const FloatType fromFloatType, const FloatType toFloatType) {
+  return isQuarkGeneratedCastOp(op, toFloatType, fromFloatType);
+}
+
+bool isOutputCastOp(
+    Operation *op, const FloatType fromFloatType, const FloatType toFloatType) {
+  return isQuarkGeneratedCastOp(op, fromFloatType, toFloatType);
+}
+
+llvm::APFloat convertF32ToBf16(
+    const llvm::fltSemantics &bf16Semantics, uint32_t f32Value) {
+  return llvm::APFloat(bf16Semantics,
+      llvm::APInt(sizeof(uint16_t) * 8, static_cast<uint16_t>(f32Value >> 16)));
```

Comment:
```suggestion
      llvm::APInt(16, static_cast<uint16_t>(f32Value >> 16)));
```
Nit: I think the name `f32Value` is a bit unfortunate. Maybe ``f32bits`